File: src/engine/renderer/buffers/BufferLayout.java
package engine.renderer.buffers;

import java.util.ArrayList;
import java.util.Arrays;

public class BufferLayout {
    private final ArrayList<BufferElement> elements;
    private int stride;

    public BufferLayout(BufferElement[] elements) {
        this.elements = new ArrayList<>(Arrays.asList(elements));
        calculateOffsetsAndStride();
    }

    public void calculateOffsetsAndStride() {
        int offset = 0;
        stride = 0;
        for (BufferElement element : elements) {
            element.offset = offset;
            offset += element.size;
            stride += (element.size * element.type.getElements());
        }
    }

    public ArrayList<BufferElement> getElements() {
        return this.elements;
    }

    public int getStride() {
        return this.stride;
    }
}

File: Lamp/Rendering/Buffers/BufferLayout.cs
using System.Collections.Generic;

namespace Lamp.Rendering.Buffers
{
    public class BufferLayout
    {
        private readonly List<BufferElement> elements;
        private int stride;

        public BufferLayout(BufferElement[] elements)
        {
            this.elements = new List<BufferElement>(elements);
            CalculateOffsetsAndStride();
        }

        public void CalculateOffsetsAndStride()
        {
            int offset = 0;
            stride = 0;
            for (int i = 0; i < elements.Count; i++)
            {
                BufferElement element = elements[i];
                int size = element.size * element.type.GetElements();
                element.offset = offset;
                offset += size;
                stride += size;
                elements[i] = element;
            }
        }

        public List<BufferElement> GetElements()
        {
            return elements;
        }

        public int GetStride()
        {
            return stride;
        }
    }
}
